qiskit/transpiler/_passmanager.py

```diff
@@ -18,14 +18,34 @@ class PassManager():
     do this correctly & efficiently
     (i.e. keep track of dependencies between passes)
     """
-    def __init__(self):
-        """Initialize an empty PassManager object
-        (with no passes scheduled).
+
+    def __init__(self, basis_gates=None, initial_layout=None):
+        """
+        Initialize an empty PassManager object (with no passes scheduled).
+
+        Args:
+            basis_gates (str): basis gates for the quantum circuit.
+            initial_layout (dict): A mapping of qubit to qubit::
+
+                              {
+                                ("q", start(int)): ("q", final(int)),
+                                ...
+                              }
+                              eg.
+                              {
+                                ("q", 0): ("q", 0),
+                                ("q", 1): ("q", 1),
+                                ("q", 2): ("q", 2),
+                                ("q", 3): ("q", 3)
+                              }
         """
         self._passes = []
+        self.shared_memory = {'basis': basis_gates.split(',') if basis_gates else [],
```

Comment:
`basis_gates` should not be part of `shared_memory`, because it is the basis_gates that all of the transpiler pipeline works towards. It is not something that gets updated with every pass. The shared memory should only contain global context about the *current* state of the DAG as the pass manager steps through different passes.

The `basis_gates` should instead be passed to the `__init()__` of different passes, if they need them. For example the `unroller` and `swap_mapper` passes may need it, but the `direction_mapper` pass does not.